[add][m]: create class FileInterface to be able to open from uploaded…

mariorodeghiero opened the pull request:
This is a draft PR to merge into branch **"feature/documentation"**

How to test in the browser:

Install dependencies :

```bash
$ yarn install
```

Create a bundle:

```bash
$ yarn build
```

create a HTML file with:

```
<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <title>Document</title>
</head>
<body>
    <div>
      <label>File:</label>
      <input type="file" id="inputField"/>
      <button>Submit file</button>
    </div>
  
   
</body>
<script src="./dist/datajs.var.js"></script>
  <script>

    var inputElement = document.getElementById("inputField");
    inputElement.addEventListener("change", handleFiles, false);
    
    async function handleFiles() {
      var fileList = this.files;
      const file = new data.open(fileList[0])
 
      await file.addSchema()
      console.log("descriptor: ", file);
      
      //Show the progress of the hash generator on console in devtools
      const hash = await file.hash(prog => console.log("Hash in Progress: " + prog))
      console.log("hash is Done:", hash);
      
      const contentBuffer = await file.buffer()
      console.log("contentBuffer :", contentBuffer);
    }
  
  </script>
</html>

```

mariorodeghiero committed:
[add][m]: create class FileInterface to be able to open from uploaded file - #50

mariorodeghiero committed:
[fix][s]: fix conflicts- #49

mariorodeghiero committed:
[add][m]: add FileInterface class - #49

- add class FIleInterface extends File
- add methods arraybuffer and hash
- fix conflicts

kmanaseryan commented on package.json:
```diff
@@ -38,6 +38,7 @@
   "dependencies": {
     "@babel/polyfill": "^7.4.4",
     "chardet": "^0.4.2",
+    "crypto-js": "^4.0.0",
```
Do we need this module?

mariorodeghiero replied:
@kmanaseryan I think yes because I try to use only `require('crypto)' but didn't work for a file from input to create a hash.
But if you know how to do with only crypto, we can try to improve.

```
// call content function and generate MD5 hash from this content
  async hash() {
    const content = await this.contentArrayBuffer()
    const wordArray = CryptoJS.lib.WordArray.create(content);

    const hash = CryptoJS.MD5(wordArray).toString()
    return hash;
  }
```

kmanaseryan commented on src/index.js:
```diff
@@ -26,6 +27,11 @@ const browser = require('./browser-utils/index')
 function open(pathOrDescriptor, { basePath, format } = {}) {
   let descriptor = null
+
+  if (browser.checkFileFromBrowser(pathOrDescriptor)) {
```
Just to keep the same style with the other functions, maybe it is worth to have another naming like `isFileFromBrowser`?

mariorodeghiero replied:
@kmanaseryan sure! let me improve it.

kmanaseryan commented on src/index.js:
```diff
@@ -126,6 +132,53 @@ class File {
+  // call content function and generate MD5 hash from this content
+  async hash() {
+    const content = await this.contentArrayBuffer()
+    const wordArray = CryptoJS.lib.WordArray.create(content);
```
Is there any specific reason using CryptoJS? I mean in another place `FileLocal` we use `crypto` in `hash()` function, maybe we can reuse the same code and module?

mariorodeghiero replied:
Is the same situation as the dependencies above, I tried to use the same logic using just crypto but didn't work for the file come from the input(browser).
But if you have any idea, we can try.

kmanaseryan reviewed (commented):
Just left some small comments, not a big deal, just some suggestions. The rest looks good. I've pulled it locally, did the test as described in the PR description. It worked pretty well... I have been able to attach file and see the logs in the console. Also run the tests no any new test failures.

mariorodeghiero committed:
[edit][s]: change the function name - #49

mariorodeghiero commented:
@kmanaseryan I sent another PR to review with the improvements.

mariorodeghiero committed:
[fix][s]: fix exports correct function name - #49

kmanaseryan committed:
[test][s]: Write a basic test for FileInterface - #49

kmanaseryan committed:
[tests][s]: Write tests for testing  addSchema() and stream() methods of file having browser File type - #49

mariorodeghiero committed:
[add][m]: create addSchema for FileInterface class - #49

- fix hash method to work with big files, tested with 1.9Gb
- create data to infer in the schema maximum 100 rows
- create readCSV function for browser to read and parse the data with support for big files without broken the browser
- used tableschema.infer to create the schema

mariorodeghiero commented:
@rufuspollock @kmanaseryan @anuveyatsu I added methods to the class FileInterface to be able to addSchema and generate the hash. 
- [x] Hash method: I fixed the hash method because broken the browser when the file is too large
  - **Note**: tested with 1.9Gb and it works properly 
- [x] addSchema method: I add some logics and I installed a library(19k size) to create the data to be able to use `tableschema.infer(data)` to create the schema.
   - **Note**:  Also tested with 1.9Gb and it works properly 

**Result** : You can check all the steps on the gif below
![schema](https://user-images.githubusercontent.com/24671133/92675329-8033a900-f2f5-11ea-9742-5e2eec8272ed.gif)

kmanaseryan committed:
Merge branch 'feature/open-from-uploaded-file' into feature/open-from-uploaded-file-test

kmanaseryan committed:
[test][m]: Integrate Karma and Jasmine for testing in real browser environement - #49

kmanaseryan commented on src/index.js:
```diff
@@ -105,6 +111,12 @@ class File {
   async addSchema() {
+    if (this.displayName === 'FileInterface') {
+      // tested with big file 1.9Gb 
+      this.descriptor.data = await browser.readCSV(this.descriptor); 
```
I guess for now having this is just only for CSV is enough, but should we have `addSchema()` for the other formats too?
cc: @anuveyatsu

anuveyatsu replied:
@kmanaseryan  Ideally, it should work for CSV and XLSX but it should be ok for now.

kmanaseryan commented on src/index.js:
```diff
@@ -126,6 +139,66 @@ class File {
+  // tested with big file 1.9Gb 
+  async hash(cbProgress) {
+    return new Promise((resolve, reject) => {
+      let md5 = CryptoJS.algo.MD5.create();
```
Just commenting for future improvements: This is going to increase the bundle size for browser, let's try use node built in `crypto` for browser instead of `CryptoJS`, since it's already being used in the proejct or use everywhere  `CryptoJS`

kmanaseryan reviewed (changes requested):
I've tested this with Karma below is the test result:

![image](https://user-images.githubusercontent.com/3371013/92724009-faf2d780-f37a-11ea-8510-dfdb1992a579.png)

So out of 3 tests 2 tests failed. 

1. `Error: We do not have a parser for that format: undefined` I guess in the constructor you should also specify `descriptor.format`. This error is thrown becuase it is `undefined` so not know format.
2. It says expected `file.descriptor.schema.fields[1].type` to have `'string'` instead of `'any'`. So I guess you need to check the field types in your implementation.

The branch which includes the test is:
`feature/open-from-uploaded-file-test`

kmanaseryan commented:
OK, please ignore number 2 failer, seems something is wrong in the test. Checking it now.

mariorodeghiero commented:
@kmanaseryan As I show you in the meeting I think the way that you create the file in your tests is different. You need to check how to use `new File([your_data])` to be sure you have the same result of the real file.

cc: @anuveyatsu @rufuspollock 

Result with a real file with the same data of the tests
![Screenshot from 2020-09-10 09-09-21](https://user-images.githubusercontent.com/24671133/92727408-6fab1f00-f345-11ea-85b1-b6cf08f9171b.png)

kmanaseryan committed:
[test][xs]: Remove extra row from test file - #49

kmanaseryan commented:
OK, it turns out that if we add extra row in the input file it behaves, so I think this is related to infer funciton. Any case I've removed the extra row at the end of the file input and now it works: addSchema() tests passes. To reproduce previous iissue even in the browser you can just add extra line in the sample.csv file and you can see it.

![image](https://user-images.githubusercontent.com/3371013/92729227-16fa7700-f383-11ea-8770-f158fb32ca6d.png)

![image](https://user-images.githubusercontent.com/3371013/92729256-1f52b200-f383-11ea-975e-194ee5613465.png)

mariorodeghiero commented:
**Note:** extra line works properly. But if you write spaces in the extra line the type will be "any".

anuveyatsu reviewed (approved):
I went through the code and especially `/src/index.js` module and it looks great! @mariorodeghiero @kmanaseryan really appreciate the effort 👍

anuveyatsu committed:
[ci][s]: replace travis with gh actions.

anuveyatsu committed:
[ci][xs]: fix syntax error in gh actions yml file.

anuveyatsu committed:
[ci][xs]: fix to run on every push.

anuveyatsu commented:
I'm merging the PR even the tests are failing because GH actions is unable to open Chrome browser to run browser tests. We can work on a separate PR for fixing the tests, otherwise, this one is becoming too large.
